Added second systems engineering challenge (#55)
gravitational/careers: challenges/systems/challenge-2.md

# Summary

Implement a prototype load balancer.

# Rationale

This exercise has two goals:

* It helps us to understand what to expect from you as a developer, how you

>   write production code, how you reason about API design and how you

Reviewer: Should we move "production"?

In the summary we say "implement a prototype" and then we say we're using that to evaluate how you write production code.

Writing code for a prototype can be vastly different than for a production system. Let's be clear about what we're looking for here.

Slack channel. That channel will contain the interview panel. You can ask them
about the engineering culture, work and life balance, or anything else that you
would like to learn about Teleport.

Before writing any actual code, we ask that you write a brief design document.
The design document should cover: design approach, scope, proposed API,
security considerations, CLI UX, and implementation details where appropriate.

Please submit the design document as Markdown and PR#1 to your repository to
allow us to provide you feedback on the proposed design.

A few notes about the design document:

* We expect the design document to be complete roughly within the first week.
  This is to ensure you have enough time to work on the implementation.
* Avoid writing an overly detailed design document. Two to three pages is
  sufficient.
* Avoid sending us draft design documents, it is difficult to evaluate which
  parts are draft and which parts are complete. Instead we encourage asking
  questions in Slack and sharing a design document that is ready to be reviewed.

Once the design document has been approved by two reviewers, move onto
implementation.

> Split your code submission into roughly 3-5 Pull Requests to give the team an

Reviewer: Should we update the guidance here? We said the design document is roughly a week, so if candidates follow that guidance they have only the 2nd week (roughly 5 work days) to put together 3-5 PRs. At the high end that's almost a PR a day, which may be hard for reviewers to keep up with.

Author: Good point, what you do think about reducing the recommended time for the design document?

Reviewer: That lines up with what I tell candidates. I usually say 2-3 days on the design doc is a good rule of thumb.

opportunity to review your code and provide feedback. Feel free to merge each
PR after you have two approvals.

Our team will do their best to provide a high quality review of the submitted
Pull Requests in a reasonable time frame. You are spending your time on this,
we are going to contribute our time too.

After the final submission, the interview team will assemble and vote using a
"+1, -2" anonymous voting system: +1 is submitted whenever a team member
accepts the submission, -2 otherwise.

In case of a positive result, we will connect you to our HR and recruiting
teams, who will work out the details and present an offer.

In case of a negative score result, hiring manager will contact you and share a
list of the key observations from the team that affected the result.

# Requirements

## Level 1

> ### Library

Reviewer: This is the first time we mention library. Do we actually care that candidates create a library that is separate from the CLI, or should we just let them organize their code as they see fit?

Author: You're right it doesn't need to be a library, but it felt like we needed some logical grouping for both the reviewers (for the early exit point) and for readability of the challenge (otherwise L5 becomes a giant list of bullet points).

What do you think?

Reviewer: Let me think about this a bit more.

If we want a library, we should add more info up front rather than casually mentioning it in the title. But maybe I can think of a better way to have an early exit point.

* Implement a simple request forwarder to forward requests to an upstream

### Server

* Develop a simple authorization scheme that defines what upstreams are
  available to which clients; this scheme can be statically defined in code
* Secure communication between client and server using TLS and token

>   authentication

Reviewer: Maybe move this bullet first, as authn is a prerequisite for authz.

* Accept and forward requests to upstreams using library

## Level 2

### Library

* Implement a round-robin request forwarder to forward requests to

>   upstreams

Reviewer: ```suggestion
  multiple upstreams
```

## Level 3

### Library

* Track number of connections per upstreams and implement a least connections
  request forwarder

### Server

* Develop a simple authorization scheme that defines what upstreams are
  available to which clients; this scheme can be statically defined in code
* Use mTLS authentication to have the server verify identity of the client and
  client of the server
* Accept and forward requests to upstreams using library

## Level 4

### Library

* Track number of connections per upstreams and implement a least connections
  request forwarder
* Track number of connections per client and implement a rate limiting request
  forwarder

### Server

* Develop a simple authorization scheme that defines what upstreams are
  available to which clients; this scheme can be statically defined in code
* Use mTLS authentication to have the server verify identity of the client and
  client of the server
* Accept and forward requests to upstreams using library

## Level 5

## Library

* Track number of connections per upstreams and implement a least connections
  request forwarder
* Track number of connections per client and implement a rate limiting request
  forwarder

> * Track unhealthy upstreams and implement a circuit breaking requests forwarder

Reviewer: When I think "circuit breaking" the first thing that comes to mind is client-side circuit breakers. Since we're not asking for a client, should we say something like "implement health checking and remove unhealthy upstreams" instead?

whiteboard/quiz interviews so common in the industry. It's not without the
downsides - it could take longer than traditional interviews.

[Some of the best teams use coding challenges.](https://sockpuppet.org/blog/2015/03/06/the-hiring-post/)

We appreciate your time and are looking forward to hacking on this project
together.

# Levels

There are 6 engineering levels at Teleport. It's possible to score on level 1-5
through coding challenge.

Level 6 is only for internal promotions. Check the [engineering levels
document](https://raw.githubusercontent.com/gravitational/careers/main/levels.pdf)
for more details.

# Interview process

The interview process will start with you receiving an invite to a private
Slack channel. That channel will contain the interview panel. You can ask them
about the engineering culture, work and life balance, or anything else that you
would like to learn about Teleport.

Before writing any actual code, we ask that you write a brief design document.
The design document should cover: design approach, scope, proposed API,
security considerations, CLI UX, and implementation details where appropriate.

> Please submit the design document as Markdown and PR#1 to your repository to

Author: Done.

allow us to provide you feedback on the proposed design.

A few notes about the design document:

* We expect the design document to be complete roughly within the first week.
  This is to ensure you have enough time to work on the implementation.
* Avoid writing an overly detailed design document. Two to three pages is
  sufficient.
* Avoid sending us draft design documents, it is difficult to evaluate which
  parts are draft and which parts are complete. Instead we encourage asking
  questions in Slack and sharing a design document that is ready to be reviewed.

> Once the design document has been approved by two reviewers, move onto

Reviewer 2: ```suggestion
Once the design document has been approved by two reviewers, move on to the
```

implementation.

Split your code submission into roughly 3-5 Pull Requests to give the team an
opportunity to review your code and provide feedback. Feel free to merge each
PR after you have two approvals.

Our team will do their best to provide a high quality review of the submitted
Pull Requests in a reasonable time frame. You are spending your time on this,
we are going to contribute our time too.

After the final submission, the interview team will assemble and vote using a
"+1, -2" anonymous voting system: +1 is submitted whenever a team member
accepts the submission, -2 otherwise.

In case of a positive result, we will connect you to our HR and recruiting
teams, who will work out the details and present an offer.

> In case of a negative score result, hiring manager will contact you and share a

Reviewer 2: ```suggestion
In case of a negative score result, the hiring manager will contact you and share a
```

list of the key observations from the team that affected the result.

# Requirements

## Level 1

### Library

> * Implement a simple request forwarder to forward requests to an upstream

Reviewer 2: We use the term "upstream" a lot. Do you guys think it is clear enough what that means?

Reviewer 2: Closing this - Zac already commented on it.

### Server

* Secure communication between client and server using TLS and token
  authentication
* Develop a simple authorization scheme that defines what upstreams are
  available to which clients; this scheme can be statically defined in code

> * Accept and forward requests to upstreams using library

Reviewer 2: ```suggestion
* Accept and forward requests to upstreams using the library
```

## Level 2

### Library

* Implement a round-robin request forwarder to forward requests to
  multiple upstreams

### Server

> * Use mTLS authentication to have the server verify identity of the client and
>   client of the server

Reviewer 2: Context is the server already.

```suggestion
* Use mTLS authentication to verify the client's identity
```

* Develop a simple authorization scheme that defines what upstreams are
  available to which clients; this scheme can be statically defined in code

> * Accept and forward requests to upstreams using library

Reviewer 2: ```suggestion
* Accept and forward requests to upstreams using library
* Copy upstream's response back to the client
```

## Level 3

### Library

> * Track number of connections per upstreams and implement a least connections

Reviewer 2: ```suggestion
* Track the number of connections per upstream and implement a least connections
```

* Implement a least connections request forwarder that tracks the number of

>   connections per upstream.

Reviewer: Yes, though we'd like to see them discuss this tradeoff in their design.

### Server

* Use mTLS authentication to have the server verify identity of the client and
  client of the server
* Develop a simple authorization scheme that defines what upstreams are
  available to which clients; this scheme can be statically defined in code
* Accept and forward requests to upstreams using library

## Level 4

### Library

* Implement a least connections request forwarder that tracks the number of
  connections per upstream.
* Implement a per-client connection rate limiter that tracks the number of

>   client connections.

Reviewer 3: We've had a few applicants asking whether the ask is to "limit the number of concurrent client connections" or "limit the number of client connections within some period of time". Is this an intended gotcha, or can we improve the wording to not point towards the incorrect interpretations. (Not sure if I should be more explicit since this is public to interviewees...)

Reviewer 3: True, but the phrasing "tracks the number of client connections" implies the other approach. IMO simply `Implement a per-client connection rate limiter` would be sufficient.

### Server

* Use mTLS authentication to have the server verify identity of the client and
  client of the server
* Develop a simple authorization scheme that defines what upstreams are
  available to which clients; this scheme can be statically defined in code
* Accept and forward requests to upstreams using library

> ## Level 5

Reviewer 2: What do you guys think of not repeating the same details for every level and saying something like "everything in level 4 plus..." instead. The first thing I did for the backend challenge was to highlight what was new for each level.

## Level 5

## Library

* Implement a least connections request forwarder that tracks the number of
  connections per upstream.
* Implement a per-client connection rate limiter that tracks the number of
  client connections.
* Implement a health checking request forwarder that removes unhealthy

>   upstreams.

Reviewer: Intentionally left as an exercise for the candidate.

  parts are draft and which parts are complete. Instead we encourage asking
  questions in Slack and sharing a design document that is ready to be reviewed.

Once the design document has been approved by two reviewers, move on to the
implementation.

Split your code submission into roughly 3-5 Pull Requests to give the team an
opportunity to review your code and provide feedback. Feel free to merge each
PR after you have two approvals.

Our team will do their best to provide a high quality review of the submitted
Pull Requests in a reasonable time frame. You are spending your time on this,
we are going to contribute our time too.

After the final submission, the interview team will assemble and vote using a
"+1, -2" anonymous voting system: +1 is submitted whenever a team member
accepts the submission, -2 otherwise.

In case of a positive result, we will connect you to our HR and recruiting
teams, who will work out the details and present an offer.

In case of a negative score result, the hiring manager will contact you and share a
list of the key observations from the team that affected the result.

# Requirements

## Level 1

### Library

> * Implement a simple request forwarder to forward requests to an upstream

Reviewer 4: Also, does L1 requirement imply that you can have 1 upstream?

* Accept and forward requests to upstreams using library

# Guidance

## Code and project ownership

This is a test challenge and we have no intent of using the code you've
submitted in production. This is your work, and you are free to do whatever you
feel is reasonable with it. In the scenario where you don't pass, you can open
source it with any license and use it as a portfolio project.

## Areas of focus

Teleport focuses on networking, infrastructure and security.

These are the areas we will be evaluating in the submission:

* Use consistent coding style. We follow [Go Coding Style]
  (https://github.com/golang/go/wiki/CodeReviewComments) for the Go
  language. If you are going to use a different language, please pick coding
  style guidelines and let us know what they are.
* Add a couple of high quality tests that cover happy and unhappy scenarios. Do
  not try to achieve full test coverage. This will take too long. Take two key
  components, e.g. authentication/authorization layer and networking and
  implement one or two test cases that demonstrate your approach to testing.
* Make sure builds are reproducible. Pick any vendoring/packaging system that

>   will allow us to get consistent build results.

Author: Yeah, this was from pre-Go modules days.

  find data races in the code. Avoid global state as much as possible; if using
  global state, write down a good description why it is necessary and protect it
  against data races.
* Deadlocks. When using mutexes, channels or any other synchronization
  primitives, make sure the system won't deadlock. We've seen candidates' code
  holding a mutex and making a network call without timeouts in place. Be extra
  careful with networking and sync primitives.
* Unstructured code. We've seen candidates leaving commented chunks of code,
  having one large file with all the code, not having code structure at all.
* Not communicating. Some candidates have submitted all their code to the
  master branch without raising pull requests, which does not give us the
  ability to provide feedback on the various implementation phases. We are a
  distributed team, so structured, asynchronous communication is critical to us.
* Implementing custom security algorithms/authentication schemes is always a
  bad idea unless you are a trained security researcher/engineer. It is
  definitely a bad idea for this task - try to stick to industry proven security
  methods as much as possible.

## Questions

> It is OK to ask the interview team questions. Some folks stay away from asking
> questions to avoid appearing less experienced, so we provide examples of
> questions to ask and questions we expect candidates to figure out on their own.

Reviewer 2: ```suggestion
It is okay to ask the interview team questions. Some folks stay away from asking
questions to avoid appearing less experienced, so we provide examples of
questions to ask and questions we expect candidates to figure out on their own.
```

Here is a great question to ask:

> Is it OK to pre-generate secret data and put the secrets in the repository
> for a proof of concept? I will add a note that we will auto-generate secrets
> in the future.

It demonstrates that you thought about this problem domain, recognize the trade
off and are saving you and the team time by not implementing it.

This is the question we expect candidates to figure out on their own:

> What version of Go should I use? What dependency manager should I use?

Unless specified in the requirements, pick the solution that works best for
you.

# Tools

This task should be implemented in Go or Rust and should work on 64-bit Linux
machines.

# Timing

You can split coding over a couple of weekdays or weekends and find time to ask
questions and receive feedback.

> Once you join the Slack channel, you have between 1 to 2 weeks complete the
> challenge depending on the challenge you choose.

Reviewer 2: ```suggestion
Once you join the Slack channel, you have between 1 to 2 weeks to complete the
challenge depending on the level you choose.
```

# Summary

Implement a basic load balancer. Load balancer are used to distribute network

> load across hosts.

Reviewer: Since we use the term upstream a lot in the rest of this document, how about:

"Load balancers are used to distribute network load across multiple hosts (or upstreams)."

Reviewer 2: I commented about the same thing. Yeah, I think it needs clarification. Hosts may work or "upstream host".

Author: Done.

### Server

* Secure communication between client and server using TLS and token

>   authentication

Reviewer: It's not completely obvious what we mean by "token authentication" or how it's supposed to work. Are we just expecting some hard coded tokens in the backend and clients just pass those through to the LB in some way?

Reviewer 2: I assume this meant something like JWT.

Author: Updated to say "opaque bearer token".
